### ISSUE TITLE
New API Proposal: Support for PEM/DER keys

### ISSUE THREAD

**Issue by 0xTim** [apple/swift-crypto#27]
#### Motivation:

Many keys are distributed from issuers in PEM or DER format, e.g.:

```
-----BEGIN PUBLIC KEY-----
MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQC0cOtPjzABybjzm3fCg1aCYwnx
PmjXpbCkecAWLj/CcDWEcuTZkYDiSG0zgglbbbhcV0vJQDWSv60tnlA3cjSYutAv
7FPo5Cq8FkvrdDzeacwRSxYuIq1LtYnd6I30qNaNthntjvbqyMmBulJ1mzLI+Xg/
aX4rbSL49Z3dAQn8vQIDAQAB
-----END PUBLIC KEY-----
```

Currently, there is no support in Swift Crypto to allow easy ingestion of keys in this form. Adding support for this would dramatically lower the barrier of entry to users trying to use PEM keys with Swift Crypto.

#### Importance:

In order to use any kind of PEM/DER keys at the moment in Swift Crypto, you need to manually implement some ASN.1 code to keys the certificate and convert it to a form that Swift Crypto understands (x963 etc). Swift Crypto [even does this](https://github.com/apple/swift-crypto/blob/64a1a98e47e6643e6d43d30b87a244483b51d8ad/Tests/CryptoTests/ECDH/BoringSSL/secpECDH_Runner_boring.swift#L64-L83) for public keys in it's test suite. The code there allows public keys to be converted, however private keys require significantly more work.

**Comment by Lukasa** [comment]
(As a note, I edited the above comment to replace instances of the word "certificate" with the word "key", as the text makes it clear that the feature request is for support for PEM/DER keys, not certs.)

Thanks for this proposal! I'll circulate it with the team and we'll get back to you.

**Comment by tanner0101** [comment]
+1 to this. It would be particularly useful for JWT libraries.

**Comment by kylebrowning** [comment]
+1

**Comment by proggeramlug** [comment]
@Lukasa has your circulation resulted in anything? :D I agree with Tim that this is a massive lack of convenience right now and should be added asap.

**Comment by RLovelett** [comment]
Me and a colleague are working on an implementation to read X.509 certificates from PEM or DER [in a fork](https://github.com/achansonjr/swift-crypto/tree/certificate_proposal). The branch `certificate_proposal` provides a new library called `ServerCrypto`.

On that branch you can do something like this:

```swift
let cert = Certificate(pem: """
-----BEGIN CERTIFICATE-----
...
-----END CERTIFICATE-----
""")
let displayIssuer = cert?.issuer?.map { "\($0.key.shortName ?? "")=\($0.value)" }.joined(separator: ", ") ?? ""
print(displayIssuer)
```

That will output something like `C=US, ST=California, L=San Fransokyo, O=San Fransokyo Institute of Technology, OU=Robotics Lab, CN=robots.sanfransokyo.edu`.

It can get other properties from an X.509 certificate too. Each property we have tried to tie back to a section of RFC 5280.

Our goal is to try to come up with a "working" implementation of this concept (based on BoringSSL) and then try and get it mainlined. We decided to put it in a "ServerCrypto" library because of this comment from the README.

> Secondly, if the API is judged not to meet the criteria for acceptance in general CryptoKit but is sufficiently important to have available for server use-cases, it will be merged into a Server namespace.

I think a best case scenario is that we, the community, come up with a consensus on an API for working with PEM/DER keys looks like and implement it. Then we propose it and that gets mainlined in either the "Crypto" library or some sort of "ServerCrypto" namespace. Next best is that Apple says "no thanks" and we just create a community fork. Worst case the community is left to fend for themselves and this code keeps getting created/repeated for every project that needs it.

**Comment by Lukasa** [comment]
Hey folks, just a quick update: we've definitely heard the feedback on PEM/DER support for EC keys, and it's something we're working on. Please remain patient with us, we'll provide more details here in the near future.

**Comment by l1ghthouse** [comment]
+1

Any update on this?

**Comment by Lukasa** [comment]
This shipped in 1.1.0 and iOS 14.
